ops/testing.py

```diff
@@ -15,44 +15,51 @@
 import inspect
 import pathlib
 from textwrap import dedent
+import typing
 
 from ops import charm, framework, model
 
 
+# OptionalYAML is something like metadata.yaml or actions.yaml. You can
+# pass in a file-like object or the string directly.
+OptionalYAML = typing.Optional[typing.Union[str, typing.TextIO]]
+
+
 # noinspection PyProtectedMember
 class Harness:
     """This class represents a way to build up the model that will drive a test suite.
 
     The model that is created is from the viewpoint of the charm that you are testing.
 
-    :ivar charm: The instance of Charm that is backed by this testing harness.
-    :type charm: CharmBase
-    """
-
-    def __init__(self, charm_cls, *, meta=None, actions=None):
-        """Used for testing your Charm or component implementations.
-
-        Example::
-
-            harness = Harness(MyCharm)
-            # Do initial setup here
-            relation_id = harness.add_relation('db', 'postgresql')
-            # Now instantiate the charm to see events as the model changes
-            harness.begin()
-            harness.add_relation_unit(relation_id, 'postgresql/0')
-            harness.update_relation_data(relation_id, 'postgresql/0', {'key': 'val'})
-            # Check that charm has properly handled the relation_joined event for postgresql/0
-            self.assertEqual(harness.charm. ...)
-
-        :param charm_cls: The Charm class that you'll be testing.
-        :param meta: (optional) A string or file-like object containing the contents of
+    Example::
+
+        harness = Harness(MyCharm)
+        # Do initial setup here
+        relation_id = harness.add_relation('db', 'postgresql')
+        # Now instantiate the charm to see events as the model changes
+        harness.begin()
+        harness.add_relation_unit(relation_id, 'postgresql/0')
+        harness.update_relation_data(relation_id, 'postgresql/0', {'key': 'val'})
+        # Check that charm has properly handled the relation_joined event for postgresql/0
+        self.assertEqual(harness.charm. ...)
+
+    Args:
+        charm_cls: The Charm class that you'll be testing.
+        meta: charm.CharmBase is a A string or file-like object containing the contents of
             metadata.yaml. If not supplied, we will look for a 'metadata.yaml' file in the
             parent directory of the Charm, and if not found fall back to a trivial
             'name: test-charm' metadata.
-        :param actions: (optional) A string or file-like object containing the contents of
+        actions: A string or file-like object containing the contents of
             actions.yaml. If not supplied, we will look for a 'actions.yaml' file in the
             parent directory of the Charm.
-        """
+    """
+
+    def __init__(
+            self,
+            charm_cls: typing.Type[charm.CharmBase],
+            *,
+            meta: OptionalYAML = None,
+            actions: OptionalYAML = None):
```

Comment:
Let's talk on next standup about this... this "every line for each" parameter is clear at some point, but having the signature across so many lines (wasting all that space to the right) is so ugly.

FTR, I'm not totally sold to this alternative either:

```
    def __init__(
            self, charm_cls: typing.Type[charm.CharmBase], *,
            meta: OptionalYAML = None, actions: OptionalYAML = None):
```

What about something more compact? 

```
    def __init__(
            self, charm_cls:charm.CharmBaseType, *,
            meta:OptionalYAML=None, actions:OptionalYAML=None):
```

Reply:
Flake8 is ok with the former, but it doesn't like the latter.
`__init__` is short enough that this doesn't seem too bad, either:
```
    def __init__(self, charm_cls: typing.Type[charm.CharmBase], *,
                 meta: OptionalYAML = None, actions: OptionalYAML = None):
```

I'm flexible. I think the fact that you can put all of the non optional args on one line and then all of the optional ones on the next works out nicely (and might be a strong contender for:
```
    def __init__(self, charm_cls: typing.Type[charm.CharmBase],
                 *, meta: OptionalYAML = None, actions: OptionalYAML = None):
```

However, if we *do* go with this, then you're likely going to end up with each of these cases being hand-tweaked. (How well does it fit on 2 lines, does it fit nicely on 3, etc).
Vs if you go with 'one line per arg' then you have a consistent design for each one.

Type annotations do end up consuming a lot of real estate, though this might be one of the most complicated that we have.